src/app_builder.rs

```diff
@@ -1,69 +1,241 @@
     track_added_plugins: HashMap<TypeId, PluginId>,
     /// track the currently being used plugin ([PluginId] is a stack since some plugins add other plugins creating a nest)
     // TODO: Track "Plugin"s for each thing
     track_current_plugin: PluginId,
     /// take a record of type names as we come across them for diagnostics
-    track_type_names: HashMap<TypeId, &'static str>,
-    /// unique type id to list of plugin type ids that provided a value for it it
-    track_uniques: HashMap<TypeId, Vec<PluginId>>,
-    /// unique type id to list of (plugin type id, reason string)
-    track_unique_dependencies: HashMap<TypeId, Vec<(PluginId, &'static str)>>,
-    /// update component storage type id to list of (plugin type id, reason string)
-    track_update_packed: HashMap<TypeId, Vec<(PluginId, &'static str)>>,
+    track_type_names: TypeNames,
+    signature: WorkloadSignature,
 }
 
-impl AppBuilder {
-    pub fn new() -> AppBuilder {
-        let mut app_builder = AppBuilder::empty();
-        app_builder.add_default_stages();
-        app_builder
+impl<'a> AppBuilder<'a> {
+    pub fn new(app: &App) -> AppBuilder<'_> {
+        AppBuilder::empty(app)
     }
+}
 
-    fn add_default_stages(&mut self) -> &mut Self {
-        #[cfg(feature = "startup-stages")]
-        {
-            self.add_startup_stage(startup_stage::STARTUP)
-                .add_startup_stage(startup_stage::POST_STARTUP);
-        }
+#[derive(Clone, Debug)]
+pub struct AppWorkload {
+    pub(crate) names: Vec<std::borrow::Cow<'static, str>>,
+}
+
+#[derive(Clone, Debug)]
+pub struct AppWorkloadInfo {
+    pub(crate) type_names: Blind<TypeNames>,
+    pub(crate) batch_info: Vec<info::BatchInfo>,
+    /// Self-imposed constraints declared by the workload
+    pub(crate) signature: Rc<WorkloadSignature>,
+    /// Derived from this plugin
+    pub(crate) plugin_id: Option<TypeId>,
+    /// Workload name assigned in the world
+    pub name: Cow<'static, str>,
+}
+
+#[derive(Clone)]
+pub(crate) struct Blind<T: Clone + 'static>(pub T);
 
-        self.add_stage(stage::FIRST)
-            .add_stage(stage::EVENT_UPDATE)
-            .add_stage(stage::PRE_UPDATE)
-            .add_stage(stage::UPDATE)
-            .add_stage(stage::POST_UPDATE)
-            .add_stage(stage::LAST)
+impl<T: Clone + 'static> std::fmt::Debug for Blind<T> {
+    fn fmt(&self, mut f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        write!(&mut f, "Blind<{}>", type_name::<T>())
     }
 }
 
-impl Default for AppBuilder {
-    fn default() -> Self {
-        AppBuilder::new()
+impl AppWorkload {
+    #[track_caller]
+    #[instrument(skip(app))]
+    pub fn run(&self, app: &App) {
+        for workload_name in self.names.iter() {
+            let span = trace_span!("AppWorkload::run", ?workload_name);
+            let _span = span.enter();
+            app.world.run_workload(&workload_name).unwrap();
```

Comment:
I'm curious about this `unwrap()`, it looks a bit out of place. Perhaps have a comment explaining why it's OK to panic with default message here (or why this panic won't happen)?

Reply:
I don't really know. I don't really know what additional info would help that much if it were to fail with a borrowing error or workload missing error.

Reply:
Part of the issue here is that shipyard doesn't yet have another level above workloads for its scheduler. But, if all goes smoothly, here, we might be able to contribute some of our designs to the future of shipyard